icu4c/source/common/uniset_props.cpp

```diff
@@ -220,438 +223,796 @@ UBool UnicodeSet::resemblesPattern(const UnicodeString& pattern, int32_t pos) {
+                                 UnicodeString& rebuiltPat,
+                                 uint32_t options,
+                                 UnicodeSet& (UnicodeSet::*caseClosure)(int32_t attribute),
+                                 int32_t depth,
+                                 UErrorCode &ec) {
+    clear();
 
-        // -------- Parse literal characters.  This includes both
-        // escaped chars ("\u4E01") and non-syntax characters
-        // ("a").
-
-        switch (lastItem) {
-        case 0:
-            lastItem = 1;
-            lastChar = c;
-            break;
-        case 1:
-            if (op == u'-') {
-                if (lastChar >= c) {
-                    // Don't allow redundant (a-a) or empty (b-a) ranges;
-                    // these are most likely typos.
-                    // syntaxError(chars, "Invalid range");
-                    ec = U_MALFORMED_SET;
-                    return;
-                }
-                add(lastChar, c);
-                _appendToPat(patLocal, lastChar, false);
-                patLocal.append(op);
-                _appendToPat(patLocal, c, false);
-                lastItem = 0;
-                op = 0;
-            } else {
-                add(lastChar, lastChar);
-                _appendToPat(patLocal, lastChar, false);
-                lastChar = c;
+    if (depth > MAX_DEPTH) {
+        U_UNICODESET_RETURN_WITH_PARSE_ERROR(("depth <= " + std::to_string(MAX_DEPTH)).c_str(),
+                                             ("depth = " + std::to_string(depth)).c_str(), lexer, ec);
+    }
+
+    bool isComplement = false;
+    // Whether to keep the syntax of the pattern at this level, only doing basic pretty-printing, e.g.,
+    // turn [ c - z[a]a - b ] into [c-z[a]a-b], but not into [a-z].
+    // This is true for a property query, or when there is a nested set.  Note that since we recurse,
+    // innermost sets consisting only of ranges will get simplified.
+    bool preserveSyntaxInPattern = false;
+    // A pattern that preserves the original syntax but strips spaces, normalizes escaping, etc.
+    UnicodeString prettyPrintedPattern;
+    if (lexer.lookahead().isPropertyQuery() || lexer.lookahead().isNamedElement()) {
+        // UnicodeSet ::= property-query | named-element
+        // NOTE(egg): For now, we ignore the work that the lexer did to find out where the
+        // property-query or named-element ended in order to retain the existing buggy behaviour of
+        // variables containing property queries.
+        lexer.getCharacterIterator().skipIgnored(lexer.charsOptions());
+        UnicodeSet propertyQuery;
+        propertyQuery.applyPropertyPattern(lexer.getCharacterIterator(), prettyPrintedPattern, ec);
+        U_UNICODESET_RETURN_IF_ERROR(ec);
+        // But now, we go back to our lexing and advance through the property-query or named-element as
+        // lexed.  If there was no error, the old and the new code should agree on the extent.
```

Comment:
After this PR lands, the reference to new vs. old code is no longer meaningful, is it?

Reply:
Well, the whole comment from NOTE(egg) to here is hopefully transitional; if the TC agrees, I will make a subsequent PR to drop make the behaviour of variables less buggy, and then we will no longer match the old code. Until then I think this comment is necessary to describe why this is written in the bonkers way it is.